### PR TITLE
Resolve Clippy `f16` and `f128` `unimplemented!`/`FIXME`s 

### PR DESCRIPTION
**Pull request by tgross35**
This was originally a PR against the Clippy repo, https://github.com/rust-lang/rust-clippy/pull/12950

r? @flip1995

Tracking issue: https://github.com/rust-lang/rust/issues/116909
Fixes: https://github.com/rust-lang/rust/pull/126636

### CODE REVIEW

**Comment by tgross35** [comment]
I know you suggested to include this as part of https://github.com/rust-lang/rust/pull/126608, but it may as well just get its own PR since it touches a lot of files. 

I need some help understanding https://github.com/rust-lang/rust-clippy/actions/runs/9569119207/job/26380925098?pr=12950#step:5:1055, I am not sure why it looks like it just deletes the file.

@rustbot label +F-f16_and_f128

File: src/tools/clippy/tests/ui/cast.rs [hunk 1]
```diff
@@ -17,6 +17,8 @@
     clippy::identity_op
 )]
 
+// FIXME(f16_f128): add tests once const casting is available for these types
```
Comment (flip1995) [inline src/tools/clippy/tests/ui/cast.rs:20]: Can those tests be added with #126608 being merged? If so, after merging this PR, please add those in 126608

File: src/tools/clippy/tests/ui/cast_size.rs [hunk 2]
```diff
@@ -1,6 +1,8 @@
 //@stderr-per-bitwidth
 //@no-rustfix
 
+#![feature(f128)]
+#![feature(f16)]
```
Comment (flip1995) [inline src/tools/clippy/tests/ui/cast_size.rs:5]: Shouldn't be necessary, as long as the tests are commented out.

File: src/tools/clippy/clippy_utils/src/consts.rs [hunk 3]
```diff
@@ -161,12 +167,18 @@ impl<'tcx> Hash for Constant<'tcx> {
...
+            },
             Self::F32(f) => {
                 f64::from(f).to_bits().hash(state);
             },
             Self::F64(f) => {
                 f.to_bits().hash(state);
             },
+            Self::F128(f) => {
+                f.to_bits().hash(state);
+            },
```
[the remark this answers is not in the document]
Comment (flip1995) [reply src/tools/clippy/clippy_utils/src/consts.rs:181]: I'm not sure why we go from f32 to f64 before hashing. But keeping this consistent for f16 seems reasonable :+1:

File: src/tools/clippy/tests/ui/cast.rs (hunk 1 above)
Comment (tgross35) [reply src/tools/clippy/tests/ui/cast.rs:20]: #126608 won't add casting, but it unblocks a followup PR that does add it. I'll make sure to add the Clippy tests with that one.

File: src/tools/clippy/clippy_utils/src/consts.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (tgross35) [reply src/tools/clippy/clippy_utils/src/consts.rs]: I assume it is to make sure there is a hash that depends only on the value and not the size, which , but I can't find anywhere that it is being used.

Unfortunately, I forgot that conversions to/from f16 aren't yet present on all platforms so I had to change this to just `f.to_bits().hash(state);`. I left a fixme here instead, maybe we could convert everything to f128 once that is better supported.

File: src/tools/clippy/tests/ui/cast_size.rs (hunk 2 above)
Comment (tgross35) [reply src/tools/clippy/tests/ui/cast_size.rs]: Thanks, I removed these

**Comment by tgross35** [comment]
I added a couple more FIXMEs in places where `Constant::{F32, F64}` is checked nonexhaustively but can't yet be added because it relies on math. @flip1995 to make things a bit more clear, here is my loose plan:

- Merge https://github.com/rust-lang/rust/pull/126608, which is blocked by this PR
- Create a new PR along the lines of https://github.com/rust-lang/rust/pull/126429 that does const casting. This is blocked on https://github.com/rust-lang/rust/pull/126608 and maybe also https://github.com/rust-lang/rust/pull/125016
- Fix the FIXMEs here that depend on const casting, which is most of the tests
- Once https://github.com/rust-lang/rust/pull/125016 merges it will unblock fixing everything that relies on non-const math. Even simple equality checks aren't available on all platforms without this.

File: src/tools/clippy/clippy_utils/src/consts.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (flip1995) [reply src/tools/clippy/clippy_utils/src/consts.rs]: > maybe we could convert everything to f128 once that is better supported.

Might be a good idea, yes. But I wonder how much that matters of if we should just use `f.to_bits().hash(state);` everywhere. But that's not for this PR to decide. Let's leave the FIXME here and address this later.

**Comment by flip1995** [comment]
Thanks for the summary what the current road map for this feature is!

Let's get this in to unblock you.

@bors r+

**Comment by tgross35** [comment]
Thank you for the review!

**Pull request merged**